### ISSUE TITLE
Problems in calculating  tracers with source

### ISSUE THREAD

**Issue by miaocb** [schism-dev/schism#79]
"vsource" and "msource" are used to calculate tr_el (schism_step.F90, line 7469-7487). 
According to my understanding, "vsource" should be real sources, not the sum of vsink, precipation and evaporation (schism_step.F90, 1602-1641).

**Comment by josephzhang8** [comment]
No, vsource in the code is the sum. If it's negative (net sink), tracer mass would leave the domain so msource should not be applied.

**Comment by miaocb** [comment]
Let's consider some scenarios:

(1) A grid element with one source ( 0.1m^3/s ) and one sink (-0.1m^3/s ), no precipation and evaporation. vsource=source+sink. "vsource" is zero, but the tracer can still be taken into the water, just like a bowl of water with one pipe putting ink into the bowl and one pipe taking water out of the bowl. The source and sink are in the same grid emelent,  but are separated from each other.

(2) A grid element with one source (0.1m^3/s), with evaporation (-0.1m^3/s)，no sink, no precipation. vsource=source+evaporation. "vsource" is zero, but tracers can also be taken into water because evaporation does not take tracers out.

(3) A grid element with one source (0.1m^3/s), with precipation (0.1m^3/s)，no sink, no evaporation. vsource=source+precipation=0.2m^/s.  Precipation does not take tracers into water, so "vsource"  can not be the sum of source and precipation in the calculation of tr_el.

**Comment by josephzhang8** [comment]
See below.


-Joseph

Joseph Zhang
Office: (804) 684 7466
Web: schism.wiki

**Comment by miaocb** [comment]
A simple experiment  is prepared for testing.
Main configuration:
1 element with size 10mx10m, no open boundary, depth is 5m
initial condition: temperature(20 degC), salinity(0)
1 source, no sink
vsource: 0 m^3/s
msource:  temperature (20 degC), salinity (35)
impose_net_flux=1, precipation-evaporation: 0.001kg/m^2/s

The source is 0 m^3/s, and as we know precipation does not take salinity to the water, salinity should be 0, but salinity increases with time as below. 
![图片](https://user-images.githubusercontent.com/11263432/188350775-7ad4961d-b9a8-4663-b032-80431c25e636.png)

Files of the experiment is attached below.
[coarse_prec.zip](https://github.com/schism-dev/schism/files/9486066/coarse_prec.zip)

**Comment by miaocb** [comment]
Hi Joseph, 
  I know that there may be no perfect method, but I think using vsoure <0 only for original sources is a slightly better choice.
(1) case1, 1 grid element with size 10mx1m,  the source and sink are in the same element.
Salt are not taken into the water, since vsource and vsink cancel out according to current method.
![图片](https://user-images.githubusercontent.com/11263432/189475013-a409e3ef-437b-4d1b-900e-c0ee53557405.png)

(2) case2, now refine the grid in case1,  with 4 grid elements, now the source and sink are in two elements separately. 
After refining the grid, source and sink are in two elements, salt are taken into the water. The results of case1 and case2 are not consinstent.
![图片](https://user-images.githubusercontent.com/11263432/189475113-21fb824d-1f02-4004-a351-2b0d68962085.png)

(3) case3, same to case2, except using vsoure <0 only for original sources.
The result is consistent with case2 with only 1 grid element.
![图片](https://user-images.githubusercontent.com/11263432/189475030-5fc77ccf-6d31-4a4b-8f2d-a43e7a0782a0.png)

**Comment by josephzhang8** [comment]
Hi Paul:

I’d argue that Cases 1 and 2 should NOT give same results; mesh resolution and placement of source/sinks do matter.

**Comment by josephzhang8** [comment]
Hi Paul:

I'd argue that Cases 1 and 2 should NOT give same results; mesh resolution and placement of source/sinks do matter.

-Joseph

Joseph Zhang
Office: (804) 684 7466
Web: schism.wiki
